samples/matter/common/src/matter_init.cpp

```diff
@@ -0,0 +1,230 @@
+#endif /* CONFIG_NET_L2_OPENTHREAD */
+
+#ifdef CONFIG_CHIP_OTA_REQUESTOR
+	/* OTA image confirmation must be done before the factory data init. */
+	OtaConfirmNewImage();
+#endif
+
+#ifdef CONFIG_MCUMGR_TRANSPORT_BT
+	/* Initialize DFU over SMP */
+	GetDFUOverSMP().Init();
+	GetDFUOverSMP().ConfirmNewImage();
+#endif
+
+	/* Initialize CHIP server */
+#if CONFIG_CHIP_FACTORY_DATA
+	VerifyOrReturn(sFactoryDataProvider, LOG_ERR("No valid factory data"));
+	sInitResult = sFactoryDataProvider->Init();
+	VerifyInitResultOrReturn(sInitResult, "FactoryDataProvider::Init() failed");
+
+	SetDeviceInstanceInfoProvider(sFactoryDataProvider);
+	SetDeviceAttestationCredentialsProvider(sFactoryDataProvider);
+	SetCommissionableDataProvider(sFactoryDataProvider);
+#else
+	SetDeviceInstanceInfoProvider(&DeviceInstanceInfoProviderMgrImpl());
+	SetDeviceAttestationCredentialsProvider(Examples::GetExampleDACProvider());
+#endif
+
+	VerifyOrReturn(sServerInitParams, LOG_ERR("No valid server initialization parameters"));
+	sInitResult = sServerInitParams->InitializeStaticResourcesBeforeServerInit();
+	VerifyInitResultOrReturn(sInitResult, "InitializeStaticResourcesBeforeServerInit() failed");
+
+	sInitResult = Server::GetInstance().Init(*sServerInitParams);
+	VerifyInitResultOrReturn(sInitResult, "Server::Init() failed");
+
+	ConfigurationMgr().LogDeviceConfig();
+	PrintOnboardingCodes(RendezvousInformationFlags(chip::RendezvousInformationFlag::kBLE));
+	AppFabricTableDelegate::Init();
+
+	sInitResult = PlatformMgr().AddEventHandler(reinterpret_cast<PlatformManager::EventHandlerFunct>(arg), 0);
+	VerifyInitResultOrReturn(sInitResult, "Cannot register CHIP event handler");
+
+	if (sCustomInit) {
+		sInitResult = sCustomInit();
+		VerifyInitResultOrReturn(sInitResult, "Custom initialization failed");
+	}
+}
+} // namespace
+
+/* Public API */
+namespace Nordic
+{
+namespace Matter
+{
+	CHIP_ERROR InitChipServer(PlatformManager::EventHandlerFunct chipEventHandler, Optional<ChipInitData> initData)
+	{
+		if (initData.HasValue()) {
+			auto customInitData = initData.Value();
+			sCommissioningInstance = customInitData.mNetworkingInstance;
+			sCustomInit = customInitData.mCustomInit;
+			sServerInitParams = customInitData.mServerInitParams;
+#if CONFIG_CHIP_FACTORY_DATA
+			sFactoryDataProvider = customInitData.mFactoryDataProvider;
+#endif
+		}
+
+		/* Before we schedule anything to execute in the CHIP thread, the platform memory
+		   and the stack itself must be initialized first. */
+		CHIP_ERROR err = Platform::MemoryInit();
+		VerifyInitResultOrReturnError(err, "Platform::MemoryInit() failed");
+		err = PlatformMgr().InitChipStack();
+		VerifyInitResultOrReturnError(err, "PlatformMgr().InitChipStack() failed");
+
+		/* Schedule all CHIP initializations to the CHIP thread for better synchronization. */
+		return PlatformMgr().ScheduleWork(DoInitChipServer, reinterpret_cast<intptr_t>(chipEventHandler));
```

Comment:
Hmm, ok it may make the whole initialization better synchronous. But for users, it will be confusing a lot I guess. Personally, I liked the previous way because it worked fine, and the whole inits were done one by one, and it was super clear for users. Doing it in a chip thread may confuse the user because they can just forget about calling "StartChipServer" and nothing will be initialized. :)

Reply:
But that's the point of using the documented API, isn't it? :) Currently the CHIP server also needs to be started with  `PlatformMgr().StartEventLoopTask();` but it's nowhere documented. So for me it sort of addresses the lack of Matter SDK API documentation.

Reply:
Discussed f2f.